File: samkhya-gpudb/scripts/wire_contract_test.mjs
// SPDX-License-Identifier: Apache-2.0
// End-to-end contract test shared by the standalone and primary TypeScript
// dummy transports. Uses only Node built-ins so CI never needs provider keys.

import assert from 'node:assert/strict';
import { spawn } from 'node:child_process';
import * as http from 'node:http';
import net from 'node:net';

const HOST = '127.0.0.1';

async function availablePort() {
  const server = net.createServer();
  await new Promise((resolve, reject) => {
    server.once('error', reject);
    server.listen(0, HOST, resolve);
  });
  const address = server.address();
  assert(address && typeof address === 'object');
  const { port } = address;
  await new Promise((resolve, reject) => server.close((error) => (error ? reject(error) : resolve())));
  return port;
}

async function request(baseUrl, path, init = {}) {
  const response = await fetch(`${baseUrl}${path}`, {
    ...init,
    signal: AbortSignal.timeout(2_000),
  });
  const raw = await response.text();
  const body = JSON.parse(raw);
  return { response, body, raw };
}

async function chunkedBody(baseUrl, byteLength) {
  const endpoint = new URL('/infer', baseUrl);
  return new Promise((resolve, reject) => {
    const req = http.request(
      {
        hostname: endpoint.hostname,
        port: endpoint.port,
        path: endpoint.pathname,
        method: 'POST',
        headers: {
          'content-type': 'application/json',
          'transfer-encoding': 'chunked',
        },
      },
      (res) => {
        let raw = '';
        res.setEncoding('utf8');
        res.on('data', (chunk) => { raw += chunk; });
        res.on('end', () => resolve({ status: res.statusCode, raw }));
      },
    );
    req.setTimeout(5_000, () => req.destroy(new Error('chunked request timed out')));
    req.on('error', reject);
    let remaining = byteLength;
    while (remaining > 0) {
      const size = Math.min(remaining, 1024 * 1024);
      req.write(Buffer.alloc(size, 0x20));
      remaining -= size;
    }
    req.end();
  });
}

async function waitUntilReady(baseUrl, child, diagnostics) {
  for (let attempt = 0; attempt < 80; attempt += 1) {
    if (child.exitCode !== null) {
      throw new Error(`server exited with ${child.exitCode}: ${diagnostics()}`);
    }
    try {
      const { response } = await request(baseUrl, '/health');
      if (response.status === 200) return;
    } catch {
      // The listener may not have bound yet.
    }
    await new Promise((resolve) => setTimeout(resolve, 50));
  }
  throw new Error(`server did not become ready: ${diagnostics()}`);
}

async function stop(child) {
  if (child.exitCode !== null) return;
  child.kill('SIGTERM');
  await Promise.race([
    new Promise((resolve) => child.once('exit', resolve)),
    new Promise((_, reject) => setTimeout(() => reject(new Error('server did not stop')), 5_000)),
  ]);
}

async function exercise(label, script, extraArgs = []) {
  const port = await availablePort();
  const baseUrl = `http://${HOST}:${port}`;
  const child = spawn(process.execPath, [script, '--host', HOST, '--port', String(port), ...extraArgs], {
    stdio: ['ignore', 'pipe', 'pipe'],
  });
  let output = '';
  child.stdout.on('data', (chunk) => { output += chunk.toString(); });
  child.stderr.on('data', (chunk) => { output += chunk.toString(); });

  try {
    await waitUntilReady(baseUrl, child, () => output);

    const health = await request(baseUrl, '/health');
    assert.equal(health.response.status, 200);
    assert.equal(health.body.ok, true);
    assert.equal(health.body.backend, 'dummy');

    const valid = await request(baseUrl, '/infer', {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify({ features: [1, 2, 3, 4, 5, 6, 7], baseline_estimate: 42 }),
    });
    assert.equal(valid.response.status, 200);
    assert.equal(valid.body.estimate, 42);
    if (valid.body._latency_ms !== undefined) {
      assert.equal(typeof valid.body._latency_ms, 'number');
      assert(valid.body._latency_ms >= 0);
    }

    const wrongWidth = await request(baseUrl, '/infer', {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify({ features: [1, 2], baseline_estimate: 42 }),
    });
    assert.equal(wrongWidth.response.status, 400);

    const nonNumericFeature = await request(baseUrl, '/infer', {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify({ features: [1, 2, 3, 4, 5, 6, '7'], baseline_estimate: 42 }),
    });
    assert.equal(nonNumericFeature.response.status, 400);

    const largeInteger = await request(baseUrl, '/infer', {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: '{"features":[1,2,3,4,5,6,7],"baseline_estimate":9007199254740993}',
    });
    assert.equal(largeInteger.response.status, 200);
    assert.match(largeInteger.raw, /"estimate":9007199254740993(?:[,}])/);

    const maxInteger = await request(baseUrl, '/infer', {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: '{"features":[1,2,3,4,5,6,7],"baseline_estimate":18446744073709551615}',
    });
    assert.equal(maxInteger.response.status, 200);
    assert.match(maxInteger.raw, /"estimate":18446744073709551615(?:[,}])/);

    const aboveMaxInteger = await request(baseUrl, '/infer', {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: '{"features":[1,2,3,4,5,6,7],"baseline_estimate":18446744073709551616}',
    });
    assert.equal(aboveMaxInteger.response.status, 400);

    const invalidJson = await request(baseUrl, '/infer', {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: '{',
    });
    assert.equal(invalidJson.response.status, 400);

    for (const nonObject of [null, [], 42]) {
      const response = await request(baseUrl, '/infer', {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(nonObject),
      });
      assert.equal(response.response.status, 400);
    }

    const oversizedBatch = await request(baseUrl, '/infer', {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify({ features: Array(7 * 1025).fill(0), baseline_estimate: 1 }),
    });
    assert.equal(oversizedBatch.response.status, 413);

    const boundaryBody = await chunkedBody(baseUrl, 8 * 1024 * 1024);
    assert.equal(boundaryBody.status, 400, boundaryBody.raw);

    const oversizedBody = await chunkedBody(baseUrl, 8 * 1024 * 1024 + 1);
    assert.equal(oversizedBody.status, 413, oversizedBody.raw);

    const missing = await request(baseUrl, '/missing');
    assert.equal(missing.response.status, 404);
  } finally {
    await stop(child);
  }

  process.stdout.write(`wire contract ok: ${label}\n`);
}

await exercise('standalone dummy', 'dist/llm_dummy_backend.js');
await exercise('primary dummy backend', 'dist/llm_infer_server.js', ['--backend', 'dummy']);

File: samkhya-gpudb/scripts/wire_contract_test.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0
"""End-to-end tests for the standalone and primary Python wire transports."""

from __future__ import annotations

import argparse
import http.client
import json
import socket
import subprocess
import sys
import time
import urllib.error
import urllib.request
from pathlib import Path
from typing import Optional


HOST = "127.0.0.1"
DUMMY_SCRIPT = Path(__file__).with_name("llm_dummy_backend.py")
PRIMARY_SCRIPT = Path(__file__).with_name("llm_infer_server.py")


def available_port() -> int:
    with socket.socket() as sock:
        sock.bind((HOST, 0))
        return int(sock.getsockname()[1])


def request(
    base_url: str, path: str, *, method: str = "GET", body: Optional[bytes] = None
) -> tuple[int, bytes]:
    headers = {"content-type": "application/json"} if body is not None else {}
    req = urllib.request.Request(
        f"{base_url}{path}", data=body, headers=headers, method=method
    )
    try:
        with urllib.request.urlopen(req, timeout=2) as response:
            return int(response.status), response.read()
    except urllib.error.HTTPError as exc:
        return int(exc.code), exc.read()


def post_json(base_url: str, value: object) -> tuple[int, bytes]:
    return request(
        base_url,
        "/infer",
        method="POST",
        body=json.dumps(value).encode("utf-8"),
    )


def wait_until_ready(base_url: str, process: subprocess.Popen) -> None:
    for _ in range(80):
        if process.poll() is not None:
            stderr = process.stderr.read() if process.stderr else ""
            raise RuntimeError(f"server exited with {process.returncode}: {stderr}")
        try:
            status, _ = request(base_url, "/health")
            if status == 200:
                return
        except OSError:
            pass
        time.sleep(0.05)
    raise TimeoutError("Python dummy server did not become ready")


def chunked_body(port: int, byte_length: int) -> int:
    """Send a headerless body of exactly ``byte_length`` bytes."""
    connection = http.client.HTTPConnection(HOST, port, timeout=5)
    try:
        connection.putrequest("POST", "/infer")
        connection.putheader("content-type", "application/json")
        connection.putheader("transfer-encoding", "chunked")
        connection.endheaders()
        remaining = byte_length
        while remaining:
            chunk = b" " * min(remaining, 1024 * 1024)
            encoded_len = f"{len(chunk):x}\r\n".encode("ascii")
            connection.send(encoded_len)
            connection.send(chunk)
            connection.send(b"\r\n")
            remaining -= len(chunk)
        connection.send(b"0\r\n\r\n")
        response = connection.getresponse()
        response.read()
        return int(response.status)
    finally:
        connection.close()


def malformed_chunk_size(port: int) -> int:
    """Send a signed chunk size, which is not valid HTTP chunk grammar."""
    connection = http.client.HTTPConnection(HOST, port, timeout=5)
    try:
        connection.putrequest("POST", "/infer")
        connection.putheader("content-type", "application/json")
        connection.putheader("transfer-encoding", "chunked")
        connection.endheaders()
        connection.send(b"-1\r\n")
        response = connection.getresponse()
        response.read()
        return int(response.status)
    finally:
        connection.close()


def exercise(
    script: Path,
    label: str,
    extra_args: list[str],
    *,
    test_chunked: bool,
    test_chunk_framing: bool,
) -> None:
    port = available_port()
    base_url = f"http://{HOST}:{port}"
    process = subprocess.Popen(
        [
            sys.executable,
            str(script),
            "--host",
            HOST,
            "--port",
            str(port),
            *extra_args,
        ],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
    )
    try:
        wait_until_ready(base_url, process)

        status, raw = request(base_url, "/health")
        assert status == 200
        health = json.loads(raw)
        assert health["ok"] is True and health["backend"] == "dummy"

        status, raw = post_json(
            base_url,
            {"features": [1, 2, 3, 4, 5, 6, 7], "baseline_estimate": 42},
        )
        valid = json.loads(raw)
        assert status == 200 and valid["estimate"] == 42
        if "_latency_ms" in valid:
            assert isinstance(valid["_latency_ms"], (int, float))
            assert valid["_latency_ms"] >= 0

        status, _ = post_json(
            base_url, {"features": [1, 2], "baseline_estimate": 42}
        )
        assert status == 400

        status, _ = post_json(
            base_url,
            {"features": [1, 2, 3, 4, 5, 6, "7"], "baseline_estimate": 42},
        )
        assert status == 400

        status, _ = post_json(
            base_url,
            {"features": [1, 2, 3, 4, 5, 6, 7], "baseline_estimate": 1 << 64},
        )
        assert status == 400

        status, raw = post_json(
            base_url,
            {"features": [1, 2, 3, 4, 5, 6, 7], "baseline_estimate": (1 << 64) - 1},
        )
        assert status == 200 and json.loads(raw)["estimate"] == (1 << 64) - 1

        status, _ = request(base_url, "/infer", method="POST", body=b"{")
        assert status == 400

        for non_object in (None, [], 42):
            status, _ = post_json(base_url, non_object)
            assert status == 400

        status, _ = post_json(
            base_url,
            {"features": [0] * (7 * 1025), "baseline_estimate": 1},
        )
        assert status == 413

        if test_chunked:
            assert chunked_body(port, 8 * 1024 * 1024) == 400
            assert chunked_body(port, 8 * 1024 * 1024 + 1) == 413
        if test_chunk_framing:
            assert malformed_chunk_size(port) == 400

        status, _ = request(base_url, "/missing")
        assert status == 404
    finally:
        process.terminate()
        try:
            process.wait(timeout=5)
        except subprocess.TimeoutExpired:
            process.kill()
            process.wait(timeout=5)

    print(f"wire contract ok: {label}")


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--primary",
        action="store_true",
        help="test the FastAPI primary server instead of the stdlib standalone server",
    )
    args = parser.parse_args()

    if args.primary:
        exercise(
            PRIMARY_SCRIPT,
            "Python primary dummy backend",
            ["--backend", "dummy"],
            test_chunked=True,
            test_chunk_framing=False,
        )
    else:
        exercise(
            DUMMY_SCRIPT,
            "Python standalone dummy",
            [],
            test_chunked=True,
            test_chunk_framing=True,
        )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
